core-bundle/src/Picker/UniversalPickerProvider.php

```diff
@@ -0,0 +1,267 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Picker;
+
+use Contao\Controller;
+use Doctrine\DBAL\Connection;
+use Knp\Menu\FactoryInterface;
+use Knp\Menu\ItemInterface;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+
+class UniversalPickerProvider implements PickerProviderInterface, DcaPickerProviderInterface, PickerMenuInterface
+{
+    /**
+     * @var FactoryInterface
+     */
+    private $menuFactory;
+
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    public function __construct(FactoryInterface $menuFactory, RouterInterface $router, TranslatorInterface $translator, Connection $connection)
+    {
+        $this->menuFactory = $menuFactory;
+        $this->router = $router;
+        $this->translator = $translator;
+        $this->connection = $connection;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName(): string
+    {
+        return 'universalPicker';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getUrl(PickerConfig $config): string
+    {
+        $table = $this->getTable($config->getContext());
+        $modules = $this->getModulesForTable($config->getContext());
+        $value = $config->getValue();
+
+        if (!$value) {
+            return $this->getUrlForValue($config, array_keys($modules)[0]);
+        }
+
+        // Use the first value if array to find a database record
+        $value = array_map('\intval', explode(',', $value))[0];
+
+        Controller::loadDataContainer($table);
+
+        $pid = null;
+        $ptable = $GLOBALS['TL_DCA'][$table]['config']['ptable'] ?? null;
+        $dynamicPtable = $GLOBALS['TL_DCA'][$table]['config']['dynamicPtable'] ?? false;
+
+        if ($ptable || $dynamicPtable) {
+            $qb = $this->connection->createQueryBuilder();
+            $qb->select('pid')->from($table)->where($qb->expr()->eq('id', $value));
+
+            if ($dynamicPtable) {
+                $qb->addSelect('ptable');
+            }
+
+            $data = $qb->execute()->fetch();
+
+            if (false === $data) {
+                return $this->getUrlForValue($config, array_keys($modules)[0]);
+            }
+
+            $pid = $data['pid'];
+
+            if ($dynamicPtable) {
+                $ptable = $data['ptable'] ?: $ptable;
+
+                // Backwards compatibility, assuming old data in tl_content
+                if (!$ptable) {
```

Comment:
This is the third nested if statement. Let's refactor this a bit to have more but simpler methods.

Reply:
Refactored in 319ef87f32e7fc220b67778d566e6a6e8fd91b35